### ISSUE TITLE
rfc 21: does "grow" introduce cycles in the job state machine?

### ISSUE THREAD

**Issue by garlick** [flux-framework/rfc#220]
[RFC 21](https://github.com/flux-framework/rfc/blob/master/spec_21.adoc) says
> The job state machine SHALL NOT contain cycles, to avoid ambiguity in synchronization.

In flux-framework/flux-core#2654, @dongahn said:
> Going forward, I think we will have to have cycles with grow and shrink

There was some discussion of cycles in #131 and #157 but it sounds like we decided to punt and keep it simple to get something working when RFC 21 was landed.

Some possible alternatives for grow:
1. A job cycles between RUN and SCHED
2. A job cycles between RUN and a new GROW state
3. A job remains in RUN state

Shrink doesn't seem like it needs to enter in to the discussion?  It's the "partial release" case that we need to handle anyway with a `free` event in the job eventlog.

**Comment by grondo** [comment]
To me it doesn't seem like grow/shrink are "states" of a job, but perhaps more like events?
(e.g. good point about partial release.)

Some other conditions where cycles may be introduced  (hypothetical, off the top of a my head) might include a job in SCHED state that has a new dependency or a hold introduced (though perhaps this can be handled without leaving the SCHED state), or a job in the RUN state is preempted and is put into a stopped or checkpointed state.

**Comment by garlick** [comment]
That was my feeling as well regarding grow, but maybe @milroy and @dongahn have more thoughts?

> Some other conditions where cycles may be introduced (hypothetical, off the top of a my head) might include a job in SCHED state that has a new dependency or a hold introduced (though perhaps this can be handled without leaving the SCHED state), or a job in the RUN state is preempted and is put into a stopped or checkpointed state.

Ah good thoughts.

I think dependencies have to be expressed up front in the current dependency design, although I guess the design isn't finalized.

We could handle "hold" similar to the way we handle a priority change.  In fact we should make that a must have feature for the near term since I think it's one of the tools currently in the hotline's toolbox.   I would argue that it need not affect the state of a job.  A job in DEPEND or SCHED could be held and prevented from actually running much as would happen if set to a low priority and sufficient demand at normal priority exists.

Should we think of "preempt" as a special case of shrink?  E.g. shrink to zero, but no "final free" bit in the eventlog so the job state remains in RUN and the epilog doesn't run?  Maybe I'm reaching a bit here :-)

**Comment by grondo** [comment]
Good points, and I don't disagree with anything. But one point of clarification:

> no "final free" bit in the eventlog so the job state remains in RUN and the epilog doesn't run? 

In the final design, the epilog should run *before* the execution system releases resources to the scheduler. This ensures that that any resource cleanup/checks specified in the epilog are complete before the resources are considered for allocation to other jobs. We also don't want to hold resources any longer than necessary for a job which is slow to exit on some nodes, so we don't want delay the epilog until after the "final free".

On current systems, jobs enter a COMPLETING state once the first epilog begins running. Flux jobs can stay in a RUN state, but there may need to be some indicator that a job is shrinking/exiting etc in the job listing tool since jobs stuck in COMPLETING for a long period tends to indicate a problem on one or more nodes.

**Comment by milroy** [comment]
The "grow" process I've been working on assumes two tiers of schedulers: a parent Flux instance and a child Flux instance running as a job.  When the child requires more resources, it submits a jobspec with a special grow attribute to its parent which will then perform a match allocate.

My somewhat uninformed opinion on this issue is that "grow" is still part of the child's "run" state because the growth is essentially a new job.  The difference is that the matched resources will be assigned the child's job id.  Basically there are two child state machines that are merged at their run states.

**Comment by milroy** [comment]
I agree with @garlick that "shrink" is a partial release.  I see it as occurring within the child's run state where the freed resources are communicated to the parent via a "detach."  As a general comment, I'm thinking of the child-parent communication as part of the child's run state.

**Comment by SteVwonder** [comment]
> I think dependencies have to be expressed up front in the current dependency design, although I guess the design isn't finalized.

That is correct.  We are planning on opening that up a bit so that a job would remain in `DEPEND` until a runtime event occurs, but IIUC even in that design, the job will not flip-flop between `DEPEND` and `SCHED`.  The runtime event will be a "one-way gate" that allows the job to move from `DEPEND` to `SCHED` but not backwards.

> We could handle "hold" similar to the way we handle a priority change. In fact we should make that a must have feature for the near term since I think it's one of the tools currently in the hotline's toolbox. I would argue that it need not affect the state of a job. A job in DEPEND or SCHED could be held and prevented from actually running much as would happen if set to a low priority and sufficient demand at normal priority exists.

Sounds good to me!

---

> A job remains in RUN state

I am leaning in this direction.  I like the idea of the top-level state machine being acyclic.  Could the `grow` and `shrink` events/states be put in the `exec` system's eventlog, making them sort of "sub states" of the `RUN` state.

**Comment by grondo** [comment]
> events/states be put in the exec system's eventlog, making them sort of "sub states" of the RUN state.

I was thinking the same thing. If you want to preserve the acyclic nature of the "macro states", any other minor states could just be considered sub-states of a macro state. At first I thought this was just kind of cheating, since you just punt the complexity you are trying to avoid down to the next level. However, on re-reading @garlick's original reasoning for the acyclic state machine, I've changed my mind and I kind of think the result is a bit brilliant...
